### PR TITLE
Translate atomic_fetch_sub to OpAtomicFAddEXT with negative value ope…

### PR DESCRIPTION
**Pull request by haonanya**
…rand

Signed-off-by: Haonan Yang <haonan.yang@intel.com>

### CODE REVIEW

**Comment by haonanya** [comment]
Hi, @svenvh, @[aratajew](https://github.com/aratajew), could you please have a look? Thanks very much.

File: test/transcoding/AtomicFSubEXTForOCL.ll [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+; CHECK-SPIRV: Extension "SPV_EXT_shader_atomic_float_add"
+; CHECK-SPIRV: TypeFloat [[TYPE_FLOAT_32:[0-9]+]] 32
+; CHECK-SPIRV: TypeFloat [[TYPE_FLOAT_64:[0-9]+]] 64
+
+
+; Function Attrs: convergent norecurse nounwind
+define dso_local spir_func void @test_atomic_float(float addrspace(1)* %a) local_unnamed_addr #0 {
+entry:
+  ; CHECK-SPIRV: 7 AtomicFAddEXT [[TYPE_FLOAT_32]]
+  ; CHECK-LLVM-CL20: call spir_func float @[[FLOAT_FUNC_NAME:_Z25atomic_fetch_add_explicit[[:alnum:]]+_Atomicff[a-zA-Z0-9_]+]](float addrspace(4)* %a.as, float -2.290000e+02, i32 0, i32 1)
```
Comment (svenvh) [inline test/transcoding/AtomicFSubEXTForOCL.ll:29]: Why use a regular expression for matching the function name?  Would it not be better to explicitly check the fully mangled name?

File: test/transcoding/AtomicFSubEXTForOCL.ll [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+attributes #2 = { convergent nounwind }
+
+!llvm.module.flags = !{!0}
+!opencl.ocl.version = !{!1}
+!opencl.spir.version = !{!1}
+!llvm.ident = !{!2}
+
+!0 = !{i32 1, !"wchar_size", i32 4}
+!1 = !{i32 2, i32 0}
+!2 = !{!"clang version 13.0.0 (https://github.com/llvm/llvm-project.git 94aa388f0ce0723bb15503cf41c2c15b288375b9)"}
```
Comment (svenvh) [inline test/transcoding/AtomicFSubEXTForOCL.ll:66]: Please remove `!2` and the `!llvm.ident` metadata, they're not needed.

(I'm aware there are many existing tests that still have it, but at least we can remove it from new tests).

File: test/transcoding/AtomicFSubEXTForOCL.ll [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+; CHECK-SPIRV: Capability AtomicFloat64AddEXT
+; CHECK-SPIRV: Extension "SPV_EXT_shader_atomic_float_add"
+; CHECK-SPIRV: TypeFloat [[TYPE_FLOAT_32:[0-9]+]] 32
+; CHECK-SPIRV: TypeFloat [[TYPE_FLOAT_64:[0-9]+]] 64
+
+
+; Function Attrs: convergent norecurse nounwind
+define dso_local spir_func void @test_atomic_float(float addrspace(1)* %a) local_unnamed_addr #0 {
+entry:
+  ; CHECK-SPIRV: 7 AtomicFAddEXT [[TYPE_FLOAT_32]]
```
Comment (aratajew) [inline test/transcoding/AtomicFSubEXTForOCL.ll:28]: Could you please add a check that would verify whether value is really negated in SPIRV? If `CHECK-LLVM-SPV` would start failing due to a missing negation then we can't easily point out whether the issue is in SPIRVWriter or SPIRVReader.

Comment (haonanya) [reply test/transcoding/AtomicFSubEXTForOCL.ll:28]: The spirv textual format is not defined by SPIR-V spec, I am not very familiar with it. 
The spirv text is:
`3 TypeFloat 3 32`
`3 TypeFloat 14 64`
...
`4 Constant 3 12 3278176256`
`5 Constant 14 20 0 3228884992`
I know Constant 3 is the float operand of atomic_fetch_sub( , float, ) and Constant 14 for atomic_fetch_sub( , double, ), but I have no idea how to check its value. Is there some doc to give some introduction? Thanks very much.

Comment (svenvh) [reply test/transcoding/AtomicFSubEXTForOCL.ll:28]: The textual format doesn't show floats as floats, but instead gives the binary representation in decimal form.

3278176256 (or `0xc3650000` in hex) is indeed `-229.0f`: https://float.exposed/0xc3650000 .  So I'd say it's fine to check for those decimal constants.

File: test/transcoding/AtomicFSubEXTForOCL.ll [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+target triple = "spir-unknown-unknown"
+
+; CHECK-SPIRV: Capability AtomicFloat32AddEXT
+; CHECK-SPIRV: Capability AtomicFloat64AddEXT
+; CHECK-SPIRV: Extension "SPV_EXT_shader_atomic_float_add"
+; CHECK-SPIRV: TypeFloat [[TYPE_FLOAT_32:[0-9]+]] 32
+; CHECK-SPIRV: TypeFloat [[TYPE_FLOAT_64:[0-9]+]] 64
+;; Check float operand of atomic_fetch_sub is handled correctly
+; CHECK-SPIRV: Constant [[TYPE_FLOAT_32]] 12 3278176256
+; CHECK-SPIRV: Constant [[TYPE_FLOAT_64]] 20 0 3228884992
```
Comment (aratajew) [inline test/transcoding/AtomicFSubEXTForOCL.ll:25]: Can you please connect these constants with checks for `AtomicFAddEXT` by using `FileCheck String Substitution Blocks`?
```suggestion
; CHECK-SPIRV: Constant [[TYPE_FLOAT_32]] [[NEGATIVE_229:[0-9]+]] 3278176256
; CHECK-SPIRV: Constant [[TYPE_FLOAT_64]] [[NEGATIVE_332:[0-9]+]] 0 3228884992
```

Comment (aratajew) [inline test/transcoding/AtomicFSubEXTForOCL.ll:25]: Can you please connect these constants with checks for `AtomicFAddEXT` by using `FileCheck String Substitution Blocks`?
```suggestion
; CHECK-SPIRV: Constant [[TYPE_FLOAT_32]] [[NEGATIVE_229:[0-9]+]] 3278176256
; CHECK-SPIRV: Constant [[TYPE_FLOAT_64]] [[NEGATIVE_332:[0-9]+]] 0 3228884992
```

Comment (haonanya) [reply test/transcoding/AtomicFSubEXTForOCL.ll]: done.

**Review by svenvh — APPROVED** [review]

**Pull request merged**